preprocessing.py

```
@@ -1,71 +1,82 @@
+            f"(max {max_categorical_features} categories each)"
+        )
 
     if not transformers:
-        detected_dtypes = X.dtypes.astype(str).to_dict()
         raise ValueError(
-            "No usable features remain after preprocessing. "
-            f"handle_categoricals={handle_categoricals}. "
-            f"detected_dtypes={detected_dtypes}"
+            f"No usable features. handle_categoricals={handle_categoricals}, "
+            f"dtypes={X.dtypes.astype(str).to_dict()}"
         )
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)